### DISCUSSION TITLE
Merge ~ Bonus ~ entries into same game

### DISCUSSION THREAD

**Discussion by Jamiras** [RetroAchievements/RAWeb#707] [Ideas]
~ Bonus ~ achievements are achievements that are considered too hard for the base set, or require glitches. They don't count towards mastery of a set or the 400 point limit.

Currently, to earn ~ Bonus ~ achievements, the player has to patch their ROM so it generates a new MD5 and fetches the ~ Bonus ~ achievement set. Instead, ~ Bonus ~ achievements should be considered part of the main game.

Suggested implementation would be to add another flag. Right now, the flags appear to be:
* 1 - active
* 2 - core
* 4 - unofficial

So, adding an 8 - bonus would seem straightforward.

When loading a game, Core and Unofficial achievements are already fetched and separated. The ~ Bonus ~ achievements would also be fetched at the same time, and activated with the Core achievements. This would allow players to earn any missing Core achievements while they're working on the ~ Bonus ~ achievements, and wouldn't require patching their ROM.

However, to make this work, we'd have to have some way to distinguish ~ Bonus ~ achievements from Core achievements on the web page and the overlays in RA_Integration and RetroArch. Additionally, RA_Integration would have to be modified to mark achievements as ~ Bonus ~, and existing ~ Bonus ~ sets would have to be migrated into their base game.

There may be additional work on the server to determine when a user Masters a set.

**Comment by julkasalsa** [comment]
Oh yeah. Bonus sets in current form are designed as something temporary, kind of work arround. I also would like to leave bonus content open, to resolve the issue of players being annoyed with revisions. In my opinion all additional achievements should be treated differently from the "backloggery" kind of stuff (achievements for progressions and features of the game).

**Comment by kdecks** [comment]
I recall @GameDragon2k making a prototype of this, in part. (No able to earn both types at the same time)

**Comment by GameDragon2k** [comment]
The prototype still exists, but there is no reason to continue with it now. It was built on the premise that there would not be any server-side extensions to handle bonus sets.

But that isn't the case anymore, so it should be handled in a more proper manner.

**Comment by kdecks** [comment]
please note on RASuite https://github.com/RetroAchievements/RASuite/issues/74

**Comment by Jamiras** [comment]
Adding a note here, but it applies to several of the linked topics as well: Achievements in Bonus sets should not affect the player's Completion rate. Or, if it does, it should be counted as excess of 100%.

For example, if a player only plays one game and completes 100% of the base game, and 10% of the bonus set, their average completion percentage should be either 100% or 110%, not 55%.

**Comment by Thoooreau** [comment]
I'd rather not counting bonus cheevos to the completion rating, so leaving at 100%. Otherwise 200% is going to be the new 100%.

**Comment by GameDragon2k** [comment]
Originally, I thought of not including Bonus sets into any form of completion percentage and simply have the additional points be the rewarding factor for them.

I am unsure if that's still a decent route to take.

**Comment by ghost** [comment]
I think the additional Sets like Bonus, Multiplayer etc. should get their own identity/percentage ratio and Completion Badge as well because without it many player don't want to bother with those Sets and its a good oppurtunity to make some creative and challenging Achievements, unlike the usual Progression Achievements in Core based which are mostly based on Main Game and collecting Items with few additional other Achievements.

I would like to see the additional Sets getting better treatment, which could also bring new users to the Site who think the Achievements in Core are to easy or not challenging/interesting enough because its always the same Achievement types just copy pasted. However, most player do prefer the basic Progression Achievements to get their Master a Core Set Badge in their profile.

High quality Bonus/Multiplayer Sets would be more for the Pro Gamer audience/streamers etc. so additional content shouldn't just be a placeholder for difficult Achievements but a proper treatment for more creative and complex Challenges, the Toolkit is contantly improving so do the Achievements and the possibilities to create some really interesting ones.

Sharing resources without cutting off on oppurtunities, Hard Type Hacks or any Hack in general which is using the same RAM could be binded to the (original source) Game Page, which would be easier for any future Hack to make Achievements for, also easier to find on the Game List, there just needs to be one option that binds the Hack MD5 only to a specific Achievement Group to exclude SMW Core (for example) being unlocked when playing a SWM Hack, same the other way around, SMW Hack should be locked when playing Original SMW so there needs to be some sort of MD5 Firewall builld into the Game Page but in the end it would be worth it by allowing to create multiple variations of same base Game, mostly Hard Type Hacks, Level Hacks. Most Hacks do have the same RAM adresses, only few (higher quality) have changed RAM adresses.

When you build something like this then you have to see the whole picture (think big) otherwise it will be just a alternative Container for Challenges nobody plays because there is nothing to earn, which is missed oppurtunity to make RetroAchievements something really interesting in future because picking up Items really gets boring after a while and there are many players who want to be challenged and getting a reward for doing it.

**Comment by Jamiras** [comment]
#1057 allows the client to fetch the subset data along with the base game data, but how should they be activated?

1) Subsets are always active.
2) A single option that enables subsets for every set.
    - If enabled, all subsets are automatically activated for any game that supports them.
3) User has to manually enable each subset. 
    - This would require restarting the game as the subsets may not expect to be activated in the middle of play
    - Presumably, we would want to remember this setting between sessions (how would we do that in RetroArch?)
4) User can manually disable each subset.
    - Could be an extension of 1 (always active) or 2 (everything activated when opted in).
    - Wouldn't necessarily need to remember which subsets are disabled between sessions.

**Reply by slashtangent** [reply]
I think maybe having them on by default but having an option to disable them might be best.  Since the player has the ability to reset unwanted achievements having them on by default isn't that big of a deal.

It will also help avoid "I wanted x subset achievements but I forgot to activate them". This way if they deactivate them it's on them.

**Reply by luchaos** [reply]
3 or 4.

I would've proposed going with opt-in in v2 but activating all subsets when first starting a game and then allow to opt-out leads to the same result.

Reasoning for me being that in v2 multiple new versions of a trigger (achievement, rp, leaderboard) may be developed by multiple people. collaboration > claims.
To make that work it would create intermediate "developer sets" which can be opted into on a user level - overriding existing compositions.

Going farther it would allow to "create your own set" by mix-matching achievements to ones own personal sets while completion will still be only ever be bound to how subsets are defined on the lowest level.

I'd go for opt-in or opt-out on the site-level in any case though. I wouldn't change much on the client side either.

**Reply by Sanaki** [reply]
I'd actually go 1 or 2, with all enabled by default. If a user doesn't want a subset, they can reset it. They can't do the opposite. Once this is normalized, a good number of normal things will end up in subsets as well, such as all fishing achievements in an RPG. It'd be a terrible precedent to have subsets disabled by default. I'm wary of even providing the option of disabling them at all.

Of note, RetroAchievements/RAIntegration#790 would do some enabling and disabling as part of that functionality, but that's also an edge case. It's something you wouldn't want to have to enable subsets to use though, if that does get implemented.

**Reply by Jamiras** [reply]
> To make that work it would create intermediate "developer sets" which can be opted into on a user level - overriding existing compositions.

There's probably still merit in something like this, but I see them more as alternate sets than subsets. Alternate sets could still be manually selected by the user. Alternate sets might even have their own subsets.

Having subsets as part of the main set is kind of a stepping stone towards #718, where each subset would award its own badge (instead of modifying the game badge - though we could change that in the future).

More importantly, it would allow developers to create "full content" achievements (i.e. "collect all rare items", or "all fishing achievements" as suggested by @Sanaki) without feeling like they're putting too much pressure on the average player.

Most importantly, it allows players to get credit for "bonus" content while playing the base set.

> Going farther it would allow to "create your own set" by mix-matching achievements to ones own personal sets while completion will still be only ever be bound to how subsets are defined on the lowest level.

Create your own set is described in #667, but as an admin feature, rather than a player feature. It would also be a framework for the AotW tracker as described in [#665](https://github.com/RetroAchievements/RAWeb/discussions/665#discussioncomment-1127171).

> I'd go for opt-in or opt-out on the site-level in any case though. I wouldn't change much on the client side either.

That's definitely an option I hadn't considered. It would make the setting be user-specific instead of client-specific. It would imply that the subset data would be sent to clients that didn't support it, but that shouldn't be a problem.

> I'm wary of even providing the option of disabling them at all.

For achievements, I completely agree. There's no real reason to not want to have subset achievements active as long as they're an extended part of the core set (and don't affect the user's completion ratio). We probably wouldn't want multiple alternative sets active at the same time.

We may want to provide an option to disable subset leaderboards. I'm not sure how many there are currently, but most of the ones I saw were things like "how far can you get in one life", so there would be additional notifications at the start of the game, and additional trackers on screen. Players may want the normal leaderboards, but not be interested in the challenge leaderboards.

**Reply by MrOwnership** [reply]
I would say option 2, with it enabled by default. The only reason say 2 over 1 is due to how subsets currently work where they are essentially their own set rather than an extension off the core set. They have their own site award, LBs, RP, unique hash and affect the average completion ration just as much as any other set does. We are going to have users who have no interest in the subsets and will have to constantly reset those subsets achievements if the option is always enabled.

When something similar to #718 or #744 (where each of the small icons represents a subset completion) does get implemented then I think switching to option 1 would be best.

A site level setting would be best, allowing the user to toggle this in their setting page.

**Reply by Tsearo** [reply]
I would prefer to have option 4 and have them all enabled by default but have the option in the emulator to disable specific ones on a per game basis with the setting saved per game. That may not be feasible though so being able to do it from the site would be fine as well imo but would still prefer to be able to specify which ones to disable. For example I may want to do the bonus subset for Pokemon but have no interest in the Professor Oak subset.

I agree that it should definitely not be forced enabled at least not in the current implementation because users shouldn't be forced to reset things in subsets they know they will never want to attempt. Also if everything was forced on there could be potential screen pollution with too many trigger icons or leaderboards for non-relevant things the user has no interest in.

Hearing things like putting fishing achievements in an RPG into a subset slightly worries me but I guess if the system is fully fleshed out in the future it will make more sense. I can see the merit in it for allowing the set to cover every possible aspect in as great of detail as needed but it's still a bit concerning without understanding the full idea. I'm just concerned what the core set would turn into in these cases with them just being the absolute basic progression if things go too far down that path with every other category of achievements being a separate subset.

Daniel brought up another issue that some subsets, particularly every professor oak challenge subset, are broken and will falsely trigger catch X pokemon achievements for players playing normally so we may need stricter guidelines for subset achievements now. For the professor oak subsets all of the catch X pokemon achievements only check the previous badges were acquired and don't check all the required pokemon from before the last badge so things like this would need to be addressed. I'm not sure if other sets will be affected but it may be worth looking into all current subsets and how they will behave when combined with normal play.

**Reply by LiquifiedSnow** [reply]
I would go with always active with the ability to turn them off if you don't want them enabled similar to hardcore. 

Like Searo mentioned, in certain scenarios we may have to have an option like in professor oaks subset not to link them to the main game for the fact that it would require rewriting all the sets not to cause any issues.

**Reply by LiquifiedSnow** [reply]
To add on as well, the ability to link a subset to multiple game's cores would be beneficial at times. I mentioned this to @Jamiras on discord DM but as an example

I'm currently working on 
https://retroachievements.org/game/19021
https://retroachievements.org/game/19012
https://retroachievements.org/game/19013
https://retroachievements.org/game/19019

and afterwards I'll be working on 
https://retroachievements.org/game/19015
https://retroachievements.org/game/19016
https://retroachievements.org/game/19017

For those not familiar you are with these games, but they have whats called "The Book of 1000". Essentially it's a giant list of stuff to do throughout all the discs, Explore x amount of maps, Kill all monsters in a dungeon x times, kill all monsters in a field x times, so on...

The first 4 are technically 1 game on 4 discs but sold separately as individual games but youre expected to transfer your data over between them same goes with the other G.U. series. Each "Volume" has it's own cap which then increases when you move your data to the next "Volume". Currently as it stands I have each volume requiring to max out in that volume in core. If a player chooses not to do it and moves on to the next Volume they have to not only finish getting all the stats from the previous volume then they also have to gather more of the stats to finish the current volume. 

Being able to link a "Book of 1000" subset to all 4 cores would make it so users arent required to grind out each volume of the game to max before moving on and could just play through the game and swap discs when ready and then finish the grind at the end of the game if they wanted to.

**Comment by Ben-Ghazi** [comment]
I'm leaning towards one but it depends on how this will affect things like completion percent or RR.

**Reply by Jamiras** [reply]
It should be easy to exclude subsets from average completion. I even suggested something like that further above.
>  Achievements in Bonus sets should not affect the player's Completion rate. Or, if it does, it should be counted as excess of 100%.

I'm not really familiar with how RR is calculated, but I imagine it could be excluded there.

**Reply by luchaos** [reply]
My interpretation was that what is now multiple games with one of them _representing_ the core set will _become_ one game with multiple _sets_.
Thus completion and points calculation remain separated. At least that'd be the easiest first pass of allowing to play multiple sets at once.

**Reply by Jamiras** [reply]
> My interpretation was that what is now multiple games with one of them representing the core set will become one game with multiple sets.

Subsets (currently) are just extensions of the base set that have things that are considered too difficult (whole game without taking damage) or too tedious (collect every 1/256 drop) for the average player.

Alternate sets is definitely something to think about in the future, but for now the goal is simply to make it possible to start earning the extended achievements while working on the main set.

**Comment by Ben-Ghazi** [comment]
If it's set up to not affect average completion, I see no reason not to go with 1.

> I'm not really familiar with how RR is calculated, but I imagine it could be excluded there.

It should be something like this: https://retroachievements.org/viewtopic.php?t=1015&c=5942. it would be great if bonus/subsets went by the player count of the base game so there's more of an incentive to get those achievements.

**Comment by redwizard42** [comment]
I don't actually play very often, but I think I'd want Bonus completion % seperate from my main completion %.

How would this work with multiple subsets?

**Comment by kdecks** [comment]
Imo ideally they should all be on by default, as most users will maintain
default settings and using subsets should be encouraged.

I'd have a in-emulator popup prompt when a subset is available that says,
"Two subsets active, to disable subsets do x". X being an easily accessible
way to switch off subsets within the emulator, and include a brief but
simply clear explanation of what subsets are.

Also have a checkbox option for subsets to be processed but muted, as in
not have any pop-ups. It might be worth considering having subsets muted by
default, if excessive pop-ups are found to be annoying.

Finally have an on-site global setting which is on by default to
disable/enable all subsets.

Hi BTW! (kdecks)

**Comment by CobaltZeroni** [comment]
I think subsets should be on by default, because "I started a subset that I didn't want to start (but I can just reset it)" is a preferable outcome to "I didn't realize subsets weren't on and missed a lot of achievements"

**Comment by FRAN269** [comment]
I think this is becoming more of a necessity now that PS2 is out, games are big in file size and subsets require another copy of the game with the patch applied, occupying double the space. Games are also bigger in terms of the content available, meaning in some if not most occasions it isn't possible to switch between core and subset to complete both simultaneously, and even then this method isn't particularly convinient. Subsets are also becoming more and more common, 19 out of 223 PS2 games have subsets created for them, it may not seem like a lot but it is considering my previous point related to file size, if both hashes could be loaded in one this would be circumvented completely.
I personally think completion percentages should be kept seperate, since merging it would mean the new full merged percentage would become the new "100%" completion percentage which isn't really the intention with this change, this is purely a quality of life improvement in my eyes, it's a matter of making it easier to enjoy these subsets and not so much about making them a part of the core experience. I also believe they should be on by default, as Cobalt said having to reset achievements you don't want is a much more preferable outcome to missing achievements you wanted. I hope this feature is still being considered and worked on because as of now the user experience is harmed by the current system and it can only get worse the more subsets are created, I honestly don't even think new pop-up notifications, percentage calculations or site changes are needed for this to have an immediate positive impact. having hashes load simultaneously alone would be a tremendous improvement to the current way of things.

**Comment by LadyLumina** [comment]
I actually made a similar post about this but never realised this was here. Silly me. Ehe. Regardless, I do believe it is also quite important due to accessibility and streamlining of RetroAchievements in general. Alongside the fact many are quite frustrated over extremely difficult achievements in Core sets. Whether it be going perfectly in the highest difficulty, flawless runs, or just outright weird tricks that a normal player would not dare attempt nor even know about given the game during its past era. Not to say it is frown upon to implement these but these all certainly belong in their own category (outside Core Mastery) due to not be directly correlated to the core design of the game to a general consumer. I suppose you can separate it as 100% completion to additional merits. One thing that immediately come to mind would be the likes of the Super Mario 64 set where most accomplishments are too trivial and obscure for their own good. They should be loaded in but outside the Core set. Which is why additional sets are pretty important. I suppose I shall leave my original post down below in case anyone is interested.

> From going through RetroAchievements for a while, I noticed quite a few potential flaws (or rather places that could use quite a few improvements), though not critical, these could be drastically improved for the ease of playing and access.
The first thing to note is the inconvenience of Subsets of games, especially in the same playthrough. This causes one to require swapping in and out from game versions just for individual achievements or at worst, do an entirely new playthrough or few just for them. These should be loaded simultaneously with the others in a separate category. For instance, going into the achievements tab on RetroArch could show the lists of types similar to how PlayStation handles DLC trophies: outside the main set (excluded from the Platinum trophy which is the equivalent of a Mastery Badge here) and the additional ones in their own category.
This might be controversial, though I would wish to bring up the frequent topic of "frustrating" achievements as expressed by quite a few. I always stand by the opinion that a single achievement can ruin an entire set, whether it is here, PlayStation, Xbox, Steam, or whatever it is. This is due to the fun factor being pushed into a category of discomfort or stress that deviates from the core experience, catering to a very small portion of the player base whilst causing passive or direct stress to other players, especially with the rising trend of full completion. As such, I would believe achievements should be allowed to be sorted (and enabled/disabled from the main to bonus sets) so the main set would usually include an experience directly correlating to a natural (full completion 100%) playthrough whilst other more hardcore sets a player would not normally partake in years ago during the original release would be sorted into another category akin to challenge and bonus sets but rather their own thing. Such as sort all the collectibles, events, stage progression, and so forth. But have the sub category available for Developers for the likes of our already known Bonus sets, and things such as no hit runs, highest difficulty playthroughs, and so forth. That way, the achievement lists will be not too daunting for players whilst Developers will have whatever freedom they want for achievement categories and requirements.
TLDR: Game files in-emulator toggle for achievement set loading. Sorting pre-existing sets (similar to the promotion and demotion of unofficial achievements) into subsets for the ease of the player base.

I do hope this can be possible one day as plenty of achievement sets suffer from the same issue but we should not limit the creativity (or insanity) of developers, letting this be a way where everyone can win.

**Reply by GabCandido** [reply]
Earlier today I was thinking about this too and I even sent a message to @wescopeland about whether I should make a post about it or not. So I'm going to take advantage of the fact that this topic is being treated here to bring my comment.

1st - The central subject of this discussion (Merge ~ Bonus ~ entries into same game) from what I understand is already a consensus and there isn't much more to be discussed, it's just a question of the viability of the code structure, thanks god. **Is there any prediction for this implementation?**

2nd - Now my comment: "I, as a player who uses RA as a guide and "proof" of fulfilling a game's content, would like to be clearer which achievements refer to the game content and which are just community challenges. I like the achievements to be categorized between "game content" and "community challenges", because, more and more, I've been losing interest in these challenges (Because many are frustrating, time consuming, and not part of the game's original intent) and would like a quick way to identify them."

Originally, I didn't think about dividing it into different sets, just categorizing it, because I imagine that a lot of annoying users could cause a stir about this, but since all the sets belong to the same game it becomes more viable in my view.

So, in short, my suggestion would be that the base set be composed of achievements that involve exploring the game's content (progress, making 100%, easter eggs, curiosities, achieve some meaningful marks in arcade games), and the so-called "community challenges", which would be any challenges created by users as an addition to the game's content/intent, being a subset. (Ex.: Get all coins in Super Mario 64/Super Mario Bros.; Do the Mercardor Challenge in Resident Evil 4, etc.)

I think it's important to make clear that I'm not against the existence of these challenges, I've only cited examples of the ones I've done, so that no one comes up with this conversation on me, I'd just like them to be an addition, something optional, something else for those who have that desire to show that they can do it. Because no, I shouldn't need to beat an entire game without dying to get some minimal recognition, being able to complete and explore all the content of a game should be enough for a "recognition badge" and anything more than that: "community challenges" "glitch showcase" "specific runs", should give extra recognition to those who are crazy enough to go after it.

I'm not talking about separating into SEVERAL categories, just separating what is exploring game's content and what is dev/community challenges.

**Reply by Xeinok-Pals** [reply]
Yeah, I agree that there seems to be a lot of confusion around this and I've written about it several times over the years: https://github.com/RetroAchievements/RAWeb/discussions/718#discussioncomment-1250583

I do like the idea of core representing in-game content and then subsets holding the out-of-game/community challenges/dev-invented challenges type of stuff, ESPECIALLY once subsets and core are playable simultaneously. That way, completionists and challenge hunters could both be more satisfied perhaps.

**Comment by Xeinok-Pals** [comment]
Maybe there is some kind of way we can have the emulators pop-up with a message on core game boot that says something like: 
"Subsets Detected:
1. Glitch Showcase
2. Low-Level Game
3. Damageless Bosses
Please enable your preferred subsets in the subset menu"

That way we can have Jamiras's Option 3 but also not the worry of people simply not knowing about subsets.

**Reply by redwizard42** [reply]
I still suspect the normal user case is "hey this game has achievements so I want to be able to earn them."

Opt-out makes more sense in this case. Present people with an option to disable them the first time they load a game, then remember this at a site level for that gamr, and give a means in-emulator and on-site to change after that.

Don't make people click extra things just to get achievements, please

**Reply by Xeinok-Pals** [reply]
I know it's a hard decision but ultimately the majority of the userbase is probably just only interested in core game completion rather than dev-created challenges/out-of-game challenges that are going to be found in subsets more, so I'm wary to disrupt the minimalist simplicity of engaging solely with a core set. Accessory lists and pop-ups for subsets might overwhelm and confuse newer users.

I think the barrier to entry of clicking a button to enable something within an emulator/on the site is far far less burdensome than looking up and patching your ROM for specific subsets which then bars you from core also though. That plus simultaneous play will be a massive QoL upgrade no matter which direction we go.

**Reply by redwizard42** [reply]
I dont want the RA experience to be worse than professional achievement systems, and in those you don't have to say 'yeah gimmie the ability to earn this games extra stuff beyond playinum'. You just can. More clicks in a UI for common tasks is bad.

The idea that "I want to be able to earn fewer achievements" is the majority opinion is odd to me.

**Reply by redwizard42** [reply]
The whole idea of this change ought to have a player not have to do anything to earn subset achievements, not make it merely less cumbersome.

**Reply by redwizard42** [reply]
Truly, my preference is to load all it can and just provide an interface somewhere to turn off earning a sunset's achievement without prompting, but a once per game upfront prompt that lets someone turn things off is a reasonable compromise to me.

**Reply by Xeinok-Pals** [reply]
Yeah, I'd be happy either way with those, I see them as only maybe like 1-2% different if the game pops up with message telling you that they are all on or all off.
